### ISSUE TITLE
LLVM 10 warning for SimCalorimetry/EcalZeroSuppressionAlgos

### ISSUE THREAD

**Issue by smuzaffar** [cms-sw/cmssw#29569]
Hi,
  @civanch and @mdhildreth ,  In latest CLANG IBs we have LLVM 10 which reveal a compilation warning for `SimCalorimetry/EcalZeroSuppressionAlgos/interface/TrivialAmplitudeAlgo.icc`  [a]. This looks like a bug to me, I think the statement should have `&&` instead of `||` here?

[a]
https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc7_amd64_gcc820/CMSSW_11_1_CLANG_X_2020-04-27-2300/SimCalorimetry/EcalZeroSuppressionProducers

```
SimCalorimetry/EcalZeroSuppressionAlgos/interface/TrivialAmplitudeAlgo.icc:34:19: warning: overlapping comparisons always evaluate to true [-Wtautological-overlap-compare]
     if (gain >= 1 || gain <= 3) {
```

**Comment by smuzaffar** [comment]
assign simulation

**Comment by smuzaffar** [comment]
I am in process of fixing other llvm warnings for simulations, so if you can suggest a proper fix then I can include that in my too.

**Comment by civanch** [comment]
@fabiocos , I suspect it is your code. It is likely correct even if the new compiler complains. May be "if" statement better rewrite in the way, which will be accepted by the compiler?

**Comment by civanch** [comment]
@fabiocos , I suspect it is your code. It is likely correct even if the new compiler complains. May be "if" statement better rewrite in the way, which will be accepted by the compiler?

**Comment by davidlange6** [comment]
if (gain >= 1 || gain <= 3) could be replaced by if (true)

**Comment by civanch** [comment]
There are two variants:

if (gain >= 1 && gain <= 3)  or if (gain <= 3)

**Comment by davidlange6** [comment]
not sure what you mean....the compiler is just telling you that

if (gain >= 1 || gain <= 3) {
      Erec = Erec + theWeights[sample] * (frame.sample(sample).adc()) * theGainFactors[gain];
    } else {
      edm::LogWarning("EcalAmplitudeError") << "Wrong gain in frame " << sample << " \n" << frame;
    }

can become

Erec = Erec + theWeights[sample] * (frame.sample(sample).adc()) * theGainFactors[gain];

if its coded correctly - which as @smuzaffar suggests it presumably is not.

**Comment by fabiocos** [comment]
@smuzaffar @civanch @davidlange6 yes, its is some very old code of mine, and it looks like "||" should have been "&&", as there are 3 possible gains only, and 0 should not be there, if I recall properly

**Comment by smuzaffar** [comment]
fix in #29597
